Computer Vision - send next largest image when full is too big

johnwatkins0 opened the pull request:
This would resolve #101 by providing backup behavior when the full-sized version of an uploaded image is greater than the maximum size accepted by Computer Vision.

Note: This does not intentionally address the "Additional Context" comment on #101 about this update potentially giving time for external image-hosting services to receive the image before the request is sent to Computer Vision. I suggest a separate issue for that. Maybe this work could be done with `wp_schedule_single_event` instead of in the same pageload as the image upload.

### Description of the Change

<!--
We must be able to understand the design of your change from this description. If we can't get a good idea of what the code will be doing from the description here, the pull request may be closed at the maintainers' discretion. Keep in mind that the maintainer reviewing this PR may not be familiar with or have worked with the code here recently, so please walk us through the concepts.  Please include screenshots (if appropriate).
-->

In the `ComputerVision` class, this adds a method to provide the URL of a newly uploaded attachment's largest acceptable version.

In most cases -- assuming most WP users don't routinely upload images greater than 4MB -- we would expect the full-size image to work, in which case this update doesn't substantively change any existing plugin behavior. When an uploaded image is greater than 4MB, however, we make a copy of the image's `size` metadata, sort the sizes from largest to smallest, and use the first size whose file is less than 4MB. 

### Alternate Designs

<!-- Explain what other alternates were considered and why the proposed version was selected. -->

A simpler approach would be to rely on the built-in core file sizes ('large', 'medium', etc.), but we don't actually know whether 'large' will be a given site's largest file size below 'full' -- and it is possible to disable core intermediate image sizes entirely -- so this approach uses a simple algorithm to sort images from largest to smallest regardless of how the sizes are named.

### Benefits

<!-- What benefits will be realized by the code change? -->

Within any site that does not have some radically altered way of handling images, Computer Vision will be able to provide data for any uploaded image.

### Possible Drawbacks

<!-- What are the possible side-effects or negative impacts of the code change? -->

Calls to the PHP `filesize` function become expensive when images are hosted externally, so the behavior implemented here should be a fallback and should not run routinely.

### Verification Process

Using a site that is reachable over the web -- which is required because Computer Vision accesses the images via a public URL -- upload an image that is larger than 4MB. After the upload is complete, the expected image metadata should have been generated.

### Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/10up/classifai/blob/develop/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my change.
- [x] All new and existing tests passed.

<!-- _NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open._ -->

### Applicable Issues

#101

adamsilverstein commented on includes/Classifai/Providers/Azure/ComputerVision.php:
```diff
@@ -81,6 +106,50 @@ public function generate_alt_tags( $metadata, $attachment_id ) {
+	/**
+	 * Retrieves the URL of the largest version of an attachment image accepted by the ComputerVision service.
+	 *
+	 * @param string $full_image The path to the full-sized image source file.
```
please keep these `@param` descriptions aligned

adamsilverstein commented:
@johnwatkins0 Overall the code looks good here.

Testing on a temporary public site however, I never saw images above 4m working. I set the threshold at 50% and small images always worked, large images never worked.

Am I missing some step to testing, are you always getting tags for larger images after your PR?

Spin up a site and test: http://poopy.life/create?plugins=https://github.com/10up/classifai/zipball/feature/handle-overlarge-images

johnwatkins0 commented:
Thanks, @adamsilverstein . I'll see if I can reproduce the issue you're having.

johnwatkins0 commented:
@adamsilverstein This is fixed now. Seems like I introduced a last-minute bug involving a function parameter. Thanks for testing!

Feel free to use this environment if you get to it in the next 22 hours -- it has the latest on this branch: http://adaptable-wren.w6.poopy.life/ I removed my image processing credentials. 

Tested with 23MB image:

<img width="1376" alt="Screen Shot 2019-09-05 at 8 52 09 PM" src="https://user-images.githubusercontent.com/9020968/64395215-0fd9b400-d01f-11e9-94f3-23afd9e64b79.png">

adamsilverstein commented:
@johnwatkins0 Thanks for fixing that up, I'll give it another test (this link should install from this branch: http://poopy.life/create?plugins=https://github.com/10up/classifai/zipball/feature/handle-overlarge-images).

Looks like some merge conflicts have cropped up, can you take a look at those?

adamsilverstein commented:
Generally this is working much better. I received some tagging results on both small and large images. Yea!

I still see an issue that I think is coming from this PR: when I drag and drop an image to upload it (vs clicking 'Select Image' I regularly see an HTTP error message. Oddly the same image uploads fine with the button select method, or sometimes when dragging and dropping, maybe if other uploads are already in progress?

Have you seen this?

![image](https://user-images.githubusercontent.com/2676022/64626309-e0be9c00-d3aa-11e9-8ca7-753ae393d3c5.png)

johnwatkins0 commented:
Thanks, @adamsilverstein . I will look into that.

adamsilverstein commented:
@johnwatkins0 Thanks! once you resolve the merge conflicts I can re-test; the issue may have been fixed elsewhere.

johnwatkins0 commented:
@adamsilverstein I've fixed the merged conflicts. I'm not able to reproduce that issue so far, but I might not be following the right steps. Let me know if you still see it.

adamsilverstein commented:
@johnwatkins0 

Steps to reproduce the HTTP error issue:

* Go to Media and Click the `Add New` button and use this interface in Media:

![image](https://user-images.githubusercontent.com/2676022/65449300-edee7880-ddf7-11e9-9746-b830592981ff.png)

* The "Add New" page also has the same issue: `/wp-admin/media-new.php`

* Drag to upload an image below 5MB in size

Note you get an HTTP error:

![image](https://user-images.githubusercontent.com/2676022/65449421-327a1400-ddf8-11e9-8e92-b753089b32a6.png)

![image](https://user-images.githubusercontent.com/2676022/65449451-3efe6c80-ddf8-11e9-9c8d-f160eed1fe45.png)


I see this notice in the logs:
`PHP Notice:  Undefined index: sizes in /plugins/classifai-for-wordpress/includes/Classifai/Providers/Azure/ComputerVision.php on line 96`

helen commented on includes/Classifai/Providers/Azure/ComputerVision.php:
```diff
@@ -60,7 +60,23 @@ public function register() {
+		 *
+		 * @param int Default 4MB.
+		 */
+		return apply_filters( 'classifai_computervision_max_filesize', 4000000 ); // 4MB default.
```
WordPress has some fun built-in constants for file sizes and time units, like `MB_IN_BYTES`. Would advise using that here for legibility and technical accuracy.

helen commented on includes/Classifai/Providers/Azure/ComputerVision.php:
```diff
@@ -74,25 +90,78 @@ public function process_image( $metadata, $attachment_id ) {
+			return $full_url;
+		}
+
+		// Sort the image sizes in order of total width + height, descending.
```
Clever!
